### PR TITLE
Bump golangci-lint to v1.42.1 and switch from golint to revive 

### PR DESCRIPTION
**Pull request by LorbusChris**
See individual commit messages for what's happening here.

/hold
til after freeze

### CODE REVIEW

**Review by cgwalters — APPROVED** [review]
LGTM, added 4.7 label.

**Comment by kikisdeliveryservice** [comment]
gonna retest first.

/retest

**Comment by LorbusChris** [comment]
/hold
I'm going to drop the third commit of this PR (which is pending the result of https://github.com/kubernetes/kubernetes/issues/96169)

**Comment by kikisdeliveryservice** [comment]
This is a lot, @runcom do we want this?

/assign @runcom

**Comment by LorbusChris** [comment]
> This is a lot, @runcom do we want this?
> 
> /assign @runcom

The second, larger commit removes one explicit TODO from the code so I do think it's desirable.

**Comment by LorbusChris** [comment]
rebased
/hold cancel

File: Makefile [hunk 1]
```diff
@@ -58,16 +61,12 @@ go-deps:
 
 install-tools:
 	GO111MODULE=on go build -o $(GOPATH)/bin/golangci-lint ./vendor/github.com/golangci/golangci-lint/cmd/golangci-lint
-	GO111MODULE=on go build -o $(GOPATH)/bin/gosec ./vendor/github.com/securego/gosec/cmd/gosec
 
 # Run verification steps
 # Example:
 #    make verify
 verify: install-tools
 	golangci-lint run --build-tags=$(GOTAGS)
```
Comment (sinnykumari) [inline Makefile:69]: does this by default invokes gosec or we need to add some flag to enable this?

File: go.mod [hunk 2]
```diff
@@ -26,27 +23,21 @@ require (
...
 	github.com/imdario/mergo v0.3.12
-	github.com/mattn/go-isatty v0.0.12 // indirect
 	github.com/opencontainers/go-digest v1.0.0
 	github.com/openshift/api v0.0.0-20210730095913-85e1d547cdee
 	github.com/openshift/client-go v0.0.0-20210521082421-73d9475a9142
 	github.com/openshift/library-go v0.0.0-20210702104503-39570b4a2ae8
 	github.com/openshift/runtime-utils v0.0.0-20210722191527-8b8348d80d1d
 	github.com/pkg/errors v0.9.1
 	github.com/prometheus/client_golang v1.7.1
-	github.com/securego/gosec v0.0.0-20191002120514-e680875ea14d
```
Comment (sinnykumari) [inline go.mod:45]: We can also remove `gosec` from replace section as well (https://github.com/openshift/machine-config-operator/blob/master/go.mod#L75)

**Comment by sinnykumari** [comment]
https://github.com/openshift/machine-config-operator/pull/2729 has been merged now. I think you can rebase and update this PR accordingly.

File: go.mod (hunk 2 above)
Comment (LorbusChris) [reply go.mod:44]: good catch, thanks! Updated

File: Makefile (hunk 1 above)
Comment (LorbusChris) [reply Makefile:69]: I've enabled it in `.golangci.yml` explicitly now, thank you!

**Comment by LorbusChris** [comment]
The newer gosec included in golanci-lint shows some interesting warnings:
``` golangci-lint run --build-tags="containers_image_openpgp exclude_graphdriver_devicemapper exclude_graphdriver_btrfs containers_image_ostree_stub"
cmd/apiserver-watcher/run.go:185:9: G306: Expect WriteFile permissions to be 0600 or less (gosec)
	err := ioutil.WriteFile(file, nil, 0644)
	       ^
test/helpers/utils.go:97:21: G404: Use of weak random number generator (math/rand instead of crypto/rand) (gosec)
	infraNode := nodes[rand.Intn(len(nodes))]
	                   ^
test/helpers/utils.go:98:14: G204: Subprocess launched with function call as argument or cmd arguments (gosec)
	out, err := exec.Command("oc", "label", "node", infraNode.Name, label+"=", "--overwrite=true").CombinedOutput()
	            ^
test/helpers/utils.go:101:14: G204: Subprocess launched with function call as argument or cmd arguments (gosec)
		out, err = exec.Command("oc", "label", "node", infraNode.Name, label+"-").CombinedOutput()
		           ^
pkg/controller/common/controller_context.go:27:13: G404: Use of weak random number generator (math/rand instead of crypto/rand) (gosec)
		factor := rand.Float64() + 1
		          ^
pkg/controller/common/helpers.go:146:2: G306: Expect WriteFile permissions to be 0600 or less (gosec)
	ioutil.WriteFile("/dev/termination-log", []byte(msg), 0644)
	^
pkg/daemon/daemon.go:174:9: G204: Subprocess launched with function call as argument or cmd arguments (gosec)
	return exec.Command("systemd-run", "--unit", "machine-config-daemon-reboot",
		"--description", fmt.Sprintf("machine-config-daemon: %s", rationale), "/bin/sh", "-c", "systemctl stop kubelet.service; systemctl reboot")
pkg/operator/bootstrap.go:204:13: G306: Expect WriteFile permissions to be 0600 or less (gosec)
		if err := ioutil.WriteFile(path, b, 0644); err != nil {
		      
```

File: .golangci.yml [hunk 3]
```diff
@@ -107,3 +107,8 @@ linters-settings:
       #- importShadow
       - paramTypeCombine
       #- unnamedResult
+  gosec:
+    severity: high
+    confidence: medium
+    excludes:
+      - G204
```
[the remark this answers is not in the document]
Comment (sinnykumari) [reply .golangci.yml:114]: Let's keep it case by case so that verify catches them for any future addition.

**Review by sinnykumari — APPROVED** [review]
Thanks for this PR!
/lgtm

**Pull request merged**